clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1480,7 +1483,8 @@ init_thread_in_process(void *drcontext)
         NOTIFY(2, "Created thread trace file %s\n", buf);
 
         /* Write initial headers at the top of the first buffer. */
-        offline_file_type_t file_type = op_L0_filter.get_value()
+        offline_file_type_t file_type =
+            op_L0_filter.get_value() || op_instr_only_trace.get_value()
             ? OFFLINE_FILE_TYPE_FILTERED
```

Comment:
I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same.  A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace.  Are we sure this ends up working properly in raw2trace?  Even if so, we may want a separate flag.

Actually I'm not really sure how a filtered no-data trace works in raw2trace: it looks like it will still go into append_memref()?  And hit the "Missing memref from predication, 0-iter repstr, or filter" case?  Is that what happens with this PR here as well (run `-verbose 4`)?

Reply:
It would be more efficient to skip the memrefs altogether.  Add a new type attribute for no data and use it to skip the memrefs?  (And set it for filtered no-data too?)

Reply:
> it looks like it will still go into append_memref()? And hit the "Missing memref from predication, 0-iter repstr, or filter" case?

Yes to both. For each memory instruction src+dst we still go into append_memref(), but return without actually appending any memref.


> I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same. A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace. Are we sure this ends up working properly in raw2trace? Even if so, we may want a separate flag.

By setting the file type to OFFLINE_FILE_TYPE_FILTERED, we're disabling address elision [here](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L773), which causes the "Missing memref..." path to trigger.

> It would be more efficient to skip the memrefs altogether. Add a new type attribute for no data and use it to skip the memrefs? (And set it for filtered no-data too?)

Do you mean by adding another field in [raw2trace_thread_data_t](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L1402) to indicate absence of memrefs?

Reply:
> > it looks like it will still go into append_memref()? And hit the "Missing memref from predication, 0-iter repstr, or filter" case?
> 
> Yes to both. For each memory instruction src+dst we still go into append_memref(), but return without actually appending any memref.
> 
> > I'm worried that conflating a filtered trace with a no-data trace will cause problems, as they are not the same. A filtered trace inserts a separate instruction entry for every insruction, while a no-data trace just has the block tags and requires reconstruction of every single instruction in raw2trace. Are we sure this ends up working properly in raw2trace? Even if so, we may want a separate flag.
> 
> By setting the file type to OFFLINE_FILE_TYPE_FILTERED, we're disabling address elision [here](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L773), which causes the "Missing memref..." path to trigger.

Right, but the worry is that a future change to raw2trace will take an action on actual filtered traces which have a very different structure from this trace and break these.

> 
> > It would be more efficient to skip the memrefs altogether. Add a new type attribute for no data and use it to skip the memrefs? (And set it for filtered no-data too?)
> 
> Do you mean by adding another field in [raw2trace_thread_data_t](https://github.com/DynamoRIO/dynamorio/blob/8551f131d97490203c2fedfbef2da9314a3ea45a/clients/drcachesim/tracer/raw2trace.h#L1402) to indicate absence of memrefs?

No, I mean adding a new offline_file_type_t to more accurately describe this: a trace with no data refs.  And using it for filtered instr-only traces too.  Then raw2trace can make better decisions internally, and we have more accurate labels.